bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/wrappers/veneer/BigtableHBaseVeneerSettings.java

```diff
@@ -218,50 +215,69 @@ public BigtableTableAdminSettings getTableAdminSettings() {
     return tableAdminSettings;
   }
 
-  /** Utility to convert {@link Configuration} to {@link BigtableInstanceAdminSettings}. */
-  @Nullable
-  public BigtableInstanceAdminSettings getInstanceAdminSettings() {
-    return instanceAdminSettings;
-  }
-
   // ************** Private Helpers **************
   private BigtableDataSettings buildBigtableDataSettings() throws IOException {
-    BigtableDataSettings.Builder dataBuilder =
-        BigtableDataSettings.newBuilder()
-            .setProjectId(getProjectId())
-            .setInstanceId(getInstanceId());
+    BigtableDataSettings.Builder dataBuilder;
+
+    // Configure the Data connection
+    Optional<String> emulatorEndpoint =
```

Comment:
please extract this out (looks like it it is repeated later on)

Reply:
I couldnt figure out a good way to extract this. The newBuilderForEmulator methods are static so there is no good way to make the helper polymorphic for TableAdminSettings and DataSettings.

I can introduce a one off value class to hold the host/port after parsing, but the combination of the extra class and separate function didnt seem worth it to replace 4 lines of splitting on ':'. It would be something like:

```java
@AutoValue
abstract class HostPort {
  abstract String getHost();
  abstract int getPort();
}
Optional<HostPort> getEmulatorEndpint() {
  Optional<String> emulatorEndpoint = Optional.fromNullable(configuration.get(BIGTABLE_EMULATOR_HOST_KEY));
  if (emulatorEndpoint.isPresent()) {
      int split = emulatorEndpoint.get().lastIndexOf(':');
      String host = emulatorEndpoint.get().substring(0, split);
      int port = Integer.parseInt(emulatorEndpoint.get().substring(split + 1));
      return new AutoValue_HostPort(host, port);
    }
return Optional.absent();
}

private BigtableDataSettings buildBigtableDataSettings() {
  Optional<HostPort> emulatorEndpoint = getEmulatorEndpint();
  if (emultorEndpoint.isPresent()) { 
     dataBuilder = BigtableDataSettings.newBuilderForEmulator(emulatorEndpoint.getHost(), emulatorEndpoint.getPort());
  } else {
    dataBuilder = BigtableDataSettings.newBuilder();
      configureConnection(dataBuilder.stubSettings(), BIGTABLE_HOST_KEY);
      configureCredentialProvider(dataBuilder.stubSettings());
 }
 ...
}

```


I guess I can push the configureCredentialProvider & configureHeaderProvider into configureConnection, but I dont think it will help much here. Any suggestions?

Reply:
Maybe a utility class? `buildBigtableDataSettings` might make more sense to leave as is, but emulator config and setup is pretty generic?